Makefile

```diff
@@ -12,130 +12,222 @@
+# TEST_IMAGE is the testing environment container image.
+TEST_IMAGE = amazon-k8s-cni-test
+TEST_IMAGE_NAME = $(TEST_IMAGE)$(IMAGE_ARCH_SUFFIX):$(VERSION)
+# These values derive ARCH and DOCKER_ARCH which are needed by dependencies in
+# image build defaulting to system's architecture when not specified.
+#
+# UNAME_ARCH is the runtime architecture of the building host.
+UNAME_ARCH = $(shell uname -m)
+# ARCH is the target architecture which is being built for.
+#
+# These are pairs of input_arch to derived_arch separated by colons:
+ARCH ?= $(lastword $(subst :, ,$(filter $(UNAME_ARCH):%,x86_64:amd64 aarch64:arm64)))
+# DOCKER_ARCH is the docker specific architecture specifier used for building on
+# multiarch container images.
+DOCKER_ARCH ?= $(lastword $(subst :, ,$(filter $(ARCH):%,amd64:amd64 arm64:arm64v8)))
+# IMAGE_ARCH_SUFFIX is the `-arch` suffix included in the container image name.
+#
+# This is only applied to the arm64 container image by default. Override to
+# provide an alternate suffix or to omit.
+IMAGE_ARCH_SUFFIX ?= $(addprefix -,$(filter $(ARCH),arm64))
+# GOLANG_IMAGE is the building golang container image used.
+GOLANG_IMAGE ?= golang:1.13-stretch
+# For the requseted build, these are the set of Go specific build environment
+# variables.
+export GOARCH = $(ARCH)
+export GOOS = linux
+export CGO_ENABLED = 0
+# NOTE: Provided for local toolchains that require explicit module feature flag.
+export GO111MODULE = on
+export GOPROXY ?= direct
+
+# LDFLAGS is the set of flags used when building golang executables.
 LDFLAGS ?= -X main.version=$(VERSION)
+# ALLPKGS is the set of packages provided in source.
+ALLPKGS = $(shell go list ./...)
+# BINS is the set of built command executables.
+BINS = aws-k8s-agent aws-cni grpc-health-probe cni-metrics-helper
+# DOCKER_ARGS is extra arguments passed during container image build.
 DOCKER_ARGS ?=
-ALLPKGS := $(shell go list ./...)
-
-ARCH ?= $(shell uname -m)
+# DOCKER_RUN_FLAGS is set the flags passed during runs of containers.
+DOCKER_RUN_FLAGS ?= --rm -ti $(DOCKER_ARGS)
+# DOCKER_BUILD_FLAGS is the set of flags passed during container image builds
+# based on the requested build.
+DOCKER_BUILD_FLAGS ?= --build-arg arch="$(ARCH)" \
+					  --build-arg docker_arch="$(DOCKER_ARCH)" \
+					  --build-arg golang_image="$(GOLANG_IMAGE)" \
+					  --network=host \
+	  		          $(DOCKER_ARGS)
+
+# Default to building an executable using the host's Go toolchaing.
+.DEFAULT_GOAL = build-linux
+
+# Build both CNI and metrics helper container images.
+all: docker docker-metrics
 
-ifeq ($(ARCH),aarch64)
-  ARCH = arm64
-else
-endif
-ifeq ($(ARCH),x86_64)
-  ARCH = amd64
-endif
+dist: all
+	mkdir -p $(DESTDIR)
+	docker save $(IMAGE_NAME) | gzip > $(IMAGE_DIST)
+	docker save $(METRICS_IMAGE_NAME) | gzip > $(METRICS_IMAGE_DIST)
 
-# Default to build the Linux binary
+# Build the VPC CNI plugin agent using host Go toolchain.
```

Comment:
s/using host Go toolchain/using the Linux Go toolchain/

..since `GOOS` is always linux, right?